### DISCUSSION TITLE
[Feature Request]: Custom nginx location config

### DISCUSSION THREAD

**Discussion by JoakimEwenson** [beyondcode/herd-community#367] [Ideas]
### Feature Description

I fairly often set up local clones of live websites, primarily WordPress sites. Lately i've been doing this in Ubuntu using custom nginx configs that use proxy locations that point requests to `/wp-content/uploads` or `/app/uploads` for Bedrock sites to the actual live page instead of making local copies of all images and other uploads.

I would love to be able to do this within Herd as well, using some form of advanced settings for individual sites. I don't need any GUI, just the option to edit the individual config files for nginx.

### Sample of custom nginx config I use
```	# Proxy for images, use /wp-content/uploads in vanilla-wp
	location /app/uploads/ {
		try_files $uri @proxy;
	}

	location @proxy {
		# Replace with actual url
		proxy_pass https://example.com;
		# Following are not required but useful for upstream server
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header X-Forwarded-Port $server_port;
	}
```

### Is this feature valuable for other users as well and why?

I do believe that it could be. I suspect more people set up local clones of live websites to do updates or changes before going to staging sites or live deploys.

**Comment by Ihabafia** [comment]
It would be great if you can either increase the timeout of nginx or have the nginx config file outside to be easer to edit.

**Comment by waterrowernohrd** [comment]
Yes please: nginx timeout is strict 60 seconds. That's not enough. Please add a custom config feature. Or has someone managed to increase the timeout?

**Comment by natebeaty** [comment]
I have some older custom sites that use nginx config for rewriting URLs, so I'd love to have this option also. I've tried editing the generated nginx configs in `~/Library/Application Support/Herd/config/valet/Nginx` without any luck, as everything seems to funnel through Valet.

**Comment by nikrowell** [comment]
I was able to load remote WordPress uploads by adding the following to `/Users/[user]/Library/Application Support/Herd/config/nginx/herd.conf`, but it's obviously not very scalable. Would love to know if there's a more efficient / dynamic way to do this.

```
location ~ ^/wp-content/uploads/(.*) {
    try_files $uri @remote;
}
location @remote {
    rewrite ^/wp-content/uploads/(.*) https://example.com/wp-content/uploads/$1 redirect;
}
```

**Comment by rabiibrahimi** [comment]
Yeah i miss this feature from laravel valet where you would create a custom nginx config file per each site and it used to work like a charm. currently using herd is limited and thinking about switching back to valet just to get that feature. i heavily work on an open source crm customization called espocrm and this is the custom nginx config file that is required to make the crm work 

```
server {
    listen 80 default_server;
    listen [::]:80 default_server;
    server_name localhost; # domain name

    charset utf-8;
    index index.html index.php;
    client_max_body_size 50M;
    keepalive_timeout 300;
    types_hash_max_size 2048;

    server_tokens off;
    fastcgi_send_timeout 300;
    fastcgi_read_timeout 300;

    gzip on;
    gzip_types text/plain text/css text/javascript application/javascript application/json;
    gzip_min_length 1000;
    gzip_comp_level 9;

    root /path-to-espo/public; # path to public dir

    location /client {
        root /path-to-espo; # path to espocrm root dir
        autoindex off;

        location ~* ^.+.(js|css|png|jpg|svg|svgz|jpeg|gif|ico|tpl)$ {
            access_log off;
            expires max;
        }
    }

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt  { access_log off; log_not_found off; }

    location ~ \.php$ {
        fastcgi_pass espocrm-php:9000;
        include fastcgi_params;
        fastcgi_index index.php;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_param QUERY_STRING $query_string;
    }

    location /api/v1/ {
        if (!-e $request_filename){
            rewrite ^/api/v1/(.*)$ /api/v1/index.php last; break;
        }
    }

    location /portal/ {
        try_files $uri $uri/ /portal/index.php?$query_string;
    }

    location /api/v1/portal-access {
        if (!-e $request_filename){
            rewrite ^/api/v1/(.*)$ /api/v1/portal-access/index.php last; break;
        }
    }

    location ~ /(\.htaccess|\web.config|\.git) {
        deny all;
    }
}
```
